tests/asgi/tests.py

```diff
@@ -333,10 +335,51 @@ def __call__(self, **kwargs):
         response_body = await communicator.receive_output()
         self.assertEqual(response_body["type"], "http.response.body")
         self.assertEqual(response_body["body"], b"Hello World!")
+        # The application gets http.disconnect as the response is complete
+        await communicator.send_input({"type": "http.disconnect"})
```

Comment:
This is the key bit. If the application server sends an `http.disonnect` before the `process_request` task completes and the `listen_for_disconnect` taks is processed first, then either we risk getting the (issue here) `request.close()` not called as `process_request` is cancelled, or it can be called twice (on different pathways) if we add that into the `try... except...` when cleaning up the tasks. 

I didn't manage to reproduce that, but it seems conceivable, yes.

Reply:
Yes its tricky to test. I think the problem is most acute where the cancellation arrives during the async code that is invoked by `await sync_to_async(request.close)()` - there isn't a easy hook to simulate that (the only async code that could be cancelled is the `sync_to_async` wrapper, not `request.close` itself as that is sync).

Having this line makes the test fail with the `await request_finished.asend()` in the `except RequestAborted` block approach that I tried last week. However, that also causes a test failure in test_asyncio_streamming_cancel_error now, so this can be safely removed I think.

The test below just sends http.disconnect in a different order, so again it fails under other patched but isn't currently being exercised, I think I can remove it.